### DISCUSSION TITLE
When I use ShaDowKHopSampler, I find the edge_index of batch exceeds the index range, is this reasonable?

### DISCUSSION THREAD

**Discussion by LCZcn96** [pyg-team/pytorch_geometric#7880] [Q&A]
After checking the source code, I found that this is caused by torch.ops.torch_sparse.ego_k_hop_sample_adj. I don't know if this is problematic. Please tell me if there is anything I am missing here, or advise me on the proper usage of ShaDowKHopSampler.

**Comment by akihironitta** [comment]
@LCZcn96 Would you mind sharing the complete error message and your code to reproduce the behaviour?

**Reply by LCZcn96** [reply]
Thanks for your replies.

I'm trying to obtain batches using ShaDowKHopSampler, with code like below.

```
for i,batch in enumerate(ShaDowKHopSampler(train_data,2,5,node_idx = None,replace=False,batch_size=128, shuffle=True)):
    print(batch.edge_index.max())
```

There are 1046 nodes in the train_data.

`Data(x=[1046, 256], edge_index=[2, 112854], edge_attr=[112854], edge_label=[56427], edge_label_index=[2, 56427])`

When doing this, I'm finding the edge_index of the batch is exceeding the index range, leading to errors.`tensor(3754)`

 I'm not sure if I'm using ShaDowKHopSampler incorrectly, or if there's something else I'm missing about how it handles batches.Thanks in advance for your time and help!

**Reply by rusty1s** [reply]
Do you have a minimal example for us to reproduce? Which dataset are you using?

**Reply by LCZcn96** [reply]
> Do you have a minimal example for us to reproduce? Which dataset are you using?

Hi,when I run the sample code [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/shadow.py), everything seems normal. 

But when I run 

```
print(data.edge_index.max())
for data in train_loader:
    print(data.edge_index.max())
```

I find the edge_index exceeds the boundary, so I realize this is my problem. 

I read the sample code, and notice that in

`model(data.x, data.edge_index, data.batch, data.root_n_id)` 

the data comes from each batch, rather than the original dataset data. This makes me feel confused, why is that? And does the edge_index in the batch correspond to the edge_index in the original dataset? Also, is there any way to get node embeddings？Thank you for your reply.

**Reply by rusty1s** [reply]
The idea is to only use the current edges in the mini-batch as input to your model. They do correspond to the original data, but are relabeled and sampled. You can obtain node embeddings by using the output of the GNN directly.

**Reply by LCZcn96** [reply]
thank you for your help ：）
